src/pools.jl

```diff
@@ -0,0 +1,109 @@
+module Pools
+
+import ..ConcurrentStack, ..Lockable
+
+export Pool, acquire, release
+
+import Base: Semaphore, acquire, release
+
+"""
+    Pool(T; max::Int=typemax(Int))
+    Pool(K, T; max::Int=typemax(Int))
+
+A threadsafe object for managing a pool of objects of type `T`, optionally keyed by objects
+of type `K`. Objects can be requested by calling `acquire(f, pool, [key])`, where `f` is a
+function that returns a new object of type `T`.
+The `key` argument is optional and can be used to lookup objects that match a certain criteria.
+If no pool exists for the given key, one will be created.
+
+If `max` is specified, the pool will limit the number of objects
+that can be checked out at any given time. If the limit has been reached, `acquire` will
+block until an object is returned to the pool via `release`.
+
+By default, `release(pool, obj)` will return the object to the pool for reuse.
+`release(pool)` will return the "permit" to the pool while not returning
+any object for reuse.
+"""
+struct Pool{K, T}
+    sem::Semaphore
+    values::Lockable{Dict{K, ConcurrentStack{T}}}
+end
+
+Pool(T; max::Int=typemax(Int)) = Pool{Nothing, T}(Semaphore(max), Lockable(Dict{Nothing, ConcurrentStack{T}}()))
```

Comment:
We might be able to do something faster in the non-key case, which I think is worth exploring (like having an extra field for a single stack that we use if the key type is `Nothing`)

Reply:
Updated timing post-refactor of the above benchmark:
```julia
julia> @btime foo()
  3.740 ms (8 allocations: 132.39 KiB)
Pool{Nothing, Int64}(Base.GenericCondition{ReentrantLock}(Base.IntrusiveLinkedList{Task}(nothing, nothing), ReentrantLock(nothing, 0x00000000, 0x00, Base.GenericCondition{Base.Threads.SpinLock}(Base.IntrusiveLinkedList{Task}(nothing, nothing), Base.Threads.SpinLock(0)), (12800, 0, 141733920768))), 10000, 0, #undef, [10001, 10002, 10003, 10004, 10005, 10006, 10007, 10008, 10009, 10010  …  19991, 19992, 19993, 19994, 19995, 19996, 19997, 19998, 19999, 20000])

```

Reply:
Love it! This is much more lean:)
I hope this is as performant in your HTTP.jl benchmarks -- the dummy benchmarks I used were not stressing any concurrent accesses. I think we're now spending more time behind lock, which originally caused you to look into lockfree datastructures.

Reply:
That was mainly because we were calling `connect` while holding the lock, which then depends on network conditions and can easily take milliseconds. I think we should be ok here. The only slight worry is that we do call `isvalid` while holding the lock, so users have to be careful that their it isn't expensive to call that (it should be fine for how we use it in HTTP)